### DISCUSSION TITLE
spectrum analyzer

### DISCUSSION THREAD

**Discussion by truckwiz** [PlummersSoftwareLLC/NightDriverStrip#656] [General]
Hi,
 I have the spectrum analyzer Night driver code working on a EZ sbc esp 32 with my own audio input, my question is how do I just choose one spectrum analyzer version? Right now it demo's all of them.

Thanks Brian

**Comment by robertlipe** [comment]
Use the web interface and disable the ones you don't like or just disable
them in the code.

Assuming you meant the project named spectrum and not generically, the
spectrum effects on, say, the project named mesmerizer, you can just whack
them in the code. There's a bug list in effect. For Spectrum, that list is
here
<https://github.com/PlummersSoftwareLLC/NightDriverStrip/blob/d2d5b05ac55d27d7c6246cf12f3a56d29b933b2f/src/effects.cpp#L402>.
The dozen different effects that the spectrum project walks through - and
thus, offers to the web interface - is defined in that big table just add
// at the front of the liens to comment them out.

Looking critically at that table, you can see there are REALLY only two
effects (The effect is the second argument, SpectrumAnalyzerEffec and
GhostWave, with the menu name as the third) so if you further wanted to
chase the remaining argument list, you can further chase those up into the
code to customize them.

If you wanted to add your own effects, that's starts pretty much in the
same blast radius of those hints. But that wasn't your question.

RJL

**Reply by truckwiz** [reply]
Thank You, I did comment out all but one, so now I either get Ghostwave demo's or SpectrumAnalyzerEffec demo's. I'm trying to not make lot's of changes to the code as that gets to be to hard to remember whats been changed.

**Comment by robertlipe** [comment]
It's not a table that gets a lot of high-volume updates, so merge
collisions shouldn't be too terrible. Almost by definition in these tables,
if there are new things, they're added to the end, so it should always be
easy to work out.
For a larger block, you can use conditional compilation to make it go away
and make that somewhat self-documenting in the process.

#if TRUCKWIZ
code()
I()
dont()
like()
#endif // TRUCKWIZ

This relies on knowing that undefined preprocessor conditionals evaluate to
zero, but then if the master copy changes dont() to do(), you won't get a
merge conflict as you would have if you whacked the line via //dont().

The other option, if you don't want to do it via the web interface, is to
use the REST API, as there's an option specifically to disable an effect:
https://github.com/PlummersSoftwareLLC/NightDriverStrip/blob/main/REST_API.md#disable-effect
Write a little script file that calls curl or postman (or newer, less
hated, alternatives) and just disables them one by one. It looks like it's
one disableEffect per effect, so it's one call per effect, but trivial to
script.

So if you're building a thousand of these and don't want to change the code
as I described or directly use the web, you can do what the web API
ultimately does anyway.

That's three options, so I think you're set!. Please mark one of these as
the accepted answer if you agree.
